### ISSUE TITLE
Reduction on GPU (numba.cuda) vs CPU yield different result

### ISSUE THREAD

**Issue by MikeQUS** [numba/numba#2480]
Why am I seeing different result on GPU compare to sequential CPU when using this reduction lambda?

```python
import numpy
from numba import cuda
from functools import reduce

A = (numpy.arange(10, dtype=numpy.float64)) + 1
cuda.reduce(lambda a, b: a + b * 20)(A) 
# result 71381.0
reduce(lambda a, b: a + b * 20, A) 
# result 1081.0
```

**Comment by seibert** [comment]
I get 1081.0 from both CPU and GPU when I run this on my test system.

Can you show the output from `numba -s`?

**Comment by MikeQUS** [comment]
```
__Hardware Information__
Machine               : x86_64
CPU Name              : skylake
CPU Features          : 
adx aes avx avx2 bmi bmi2 clflushopt cmov cx16 f16c fma fsgsbase hle invpcid
lzcnt mmx movbe pclmul popcnt prfchw rdrnd rdseed rtm sgx smap sse sse2 sse3
sse4.1 sse4.2 ssse3 xsave xsavec xsaveopt xsaves

__OS Information__
Platform              : Linux-4.4.0-83-generic-x86_64-with-debian-stretch-sid
Release               : 4.4.0-83-generic
System Name           : Linux
Version               : #106-Ubuntu SMP Mon Jun 26 17:54:43 UTC 2017
OS specific info      : debianstretch/sid
glibc info            : glibc 2.2.5

__Python Information__
Python Compiler       : GCC 4.4.7 20120313 (Red Hat 4.4.7-1)
Python Implementation : CPython
Python Version        : 2.7.13
Python Locale         : en_US UTF-8

__LLVM information__
LLVM version          : 4.0.0

__CUDA Information__
Found 1 CUDA devices
id 0     GeForce GTX 1080                               [SUPPORTED]
                      compute capability: 6.1
                           pci device id: 0
                              pci bus id: 1
Summary:
	1/1 devices are supported
CUDA driver version   : 9000
CUDA libraries:
Finding cublas
	ERROR: can't locate lib
Finding cusparse
	ERROR: can't locate lib
Finding cufft
	ERROR: can't locate lib
Finding curand
	ERROR: can't locate lib
Finding nvvm
	ERROR: can't locate lib
	finding libdevice for compute_20...	ERROR: can't open libdevice for compute_20
	finding libdevice for compute_30...	ERROR: can't open libdevice for compute_30
	finding libdevice for compute_35...	ERROR: can't open libdevice for compute_35
	finding libdevice for compute_50...	ERROR: can't open libdevice for compute_50

__Conda Information__
conda_build_version   : 3.0.8
conda_env_version     : 4.3.22
platform              : linux-64
python_version        : 2.7.13.final.0
root_writable         : True

__Current Conda Env__
_license                  1.1                      py27_1  
alabaster                 0.7.10                   py27_0  
anaconda                  4.4.0               np112py27_0  
anaconda-client           1.6.3                    py27_0  
anaconda-navigator        1.6.2                    py27_0  
anaconda-project          0.6.0                    py27_0  
argcomplete               1.0.0                    py27_1  
asn1crypto                0.22.0                   py27_0  
astroid                   1.4.9                    py27_0  
astroid                   1.4.8                     <pip>
astropy                   1.3.2               np112py27_0  
asv                       0.3.dev1099+a95b68de           <pip>
babel                     2.4.0                    py27_0  
backports                 1.0                      py27_0  
backports.functools-lru-cache 1.3                       <pip>
backports_abc             0.5                      py27_0  
beautifulsoup4            4.6.0                    py27_0  
bitarray                  0.8.1                    py27_0  
blaze                     0.10.1                   py27_0  
bleach                    1.5.0                    py27_0  
bokeh                     0.12.5                   py27_1  
boto                      2.46.1                   py27_0  
bottleneck                1.2.1               np112py27_0  
cairo                     1.14.8                        0  
cdecimal                  2.3                      py27_2  
cffi                      1.10.0                   py27_0  
chardet                   3.0.3                    py27_0  
chest                     0.2.3                    py27_0  
click                     6.7                      py27_0  
cloudpickle               0.2.2                    py27_0  
clyent                    1.2.2                    py27_0  
colorama                  0.3.9                    py27_0  
conda                     4.3.22                   py27_0  
conda-build               3.0.8                    py27_0  
conda-env                 2.6.0                         0  
conda-manager             0.4.0                    py27_0  
conda-verify              2.0.0                    py27_0  
configobj                 5.0.6                    py27_0  
configparser              3.5.0                    py27_0  
configparser              3.5.0                     <pip>
contextlib2               0.5.5                    py27_0  
cryptography              1.8.1                    py27_0  
curl                      7.52.1                        0  
cycler                    0.10.0                   py27_0  
cython                    0.25.2                   py27_0  
cytoolz                   0.8.2                    py27_0  
dask                      0.14.3                   py27_1  
datashape                 0.5.4                    py27_0  
dbus                      1.10.10                       0  
decorator                 4.0.11                   py27_0  
dill                      0.2.6                    py27_0  
distributed               1.16.3                   py27_0  
docutils                  0.13.1                   py27_0  
dynd-python               0.7.2                    py27_0  
entrypoints               0.2.2                    py27_1  
enum34                    1.1.6                    py27_0  
et_xmlfile                1.0.1                    py27_0  
expat                     2.1.0                         0  
fastcache                 1.0.2                    py27_1  
filelock                  2.0.7                    py27_0  
flask                     0.12.2                   py27_0  
flask-cors                3.0.2                    py27_0  
fontconfig                2.12.1                        3  
freetype                  2.5.5                         2  
funcsigs                  1.0.2                    py27_0  
functools32               3.2.3.2                  py27_0  
futures                   3.1.1                    py27_0  
get_terminal_size         1.0.0                    py27_0  
gevent                    1.2.1                    py27_0  
glib                      2.50.2                        1  
glob2                     0.5                      py27_0  
greenlet                  0.4.12                   py27_0  
grin                      1.2.1                    py27_3  
gst-plugins-base          1.8.0                         0  
gstreamer                 1.8.0                         0  
h5py                      2.7.0               np112py27_0  
harfbuzz                  0.9.39                        2  
hdf5                      1.8.17                        1  
heapdict                  1.0.0                    py27_1  
html5lib                  0.999                    py27_0  
icu                       54.1                          0  
idna                      2.5                      py27_0  
imagesize                 0.7.1                    py27_0  
ipaddress                 1.0.18                   py27_0  
ipykernel                 4.6.1                    py27_0  
ipython                   5.3.0                    py27_0  
ipython_genutils          0.2.0                    py27_0  
ipywidgets                6.0.0                    py27_0  
isort                     4.2.5                    py27_0  
isort                     4.2.5                     <pip>
itsdangerous              0.24                     py27_0  
jbig                      2.1                           0  
jdcal                     1.3                      py27_0  
jedi                      0.10.2                   py27_2  
jinja2                    2.9.6                    py27_0  
jpeg                      9b                            0  
jsonschema                2.6.0                    py27_0  
jupyter                   1.0.0                    py27_3  
jupyter_client            5.0.1                    py27_0  
jupyter_console           5.1.0                    py27_0  
jupyter_core              4.3.0                    py27_0  
lazy-object-proxy         1.2.2                    py27_0  
lazy-object-proxy         1.2.2                     <pip>
libdynd                   0.7.2                         0  
libffi                    3.2.1                         1  
libgcc                    4.8.5                         2  
libgfortran               3.0.0                         1  
libiconv                  1.14                          0  
libpng                    1.6.27                        0  
libsodium                 1.0.10                        0  
libtiff                   4.0.6                         3  
libtool                   2.4.2                         0  
libxcb                    1.12                          1  
libxml2                   2.9.4                         0  
libxslt                   1.1.29                        0  
llvmlite                  0.18.0                   py27_0  
locket                    0.2.0                    py27_1  
lxml                      3.7.3                    py27_0  
markupsafe                0.23                     py27_2  
matplotlib                2.0.2               np112py27_0  
mccabe                    0.5.2                     <pip>
mistune                   0.7.4                    py27_0  
mkl                       2017.0.1                      0  
mkl-service               1.1.2                    py27_3  
mpmath                    0.19                     py27_1  
msgpack-python            0.4.8                    py27_0  
multipledispatch          0.4.9                    py27_0  
navigator-updater         0.1.0                    py27_0  
nbconvert                 5.1.1                    py27_0  
nbformat                  4.3.0                    py27_0  
networkx                  1.11                     py27_0  
nltk                      3.2.3                    py27_0  
nose                      1.3.7                    py27_1  
notebook                  5.0.0                    py27_0  
numba                     0.33.0              np112py27_0  
numexpr                   2.6.2               np112py27_0  
numpy                     1.12.1                   py27_0  
numpydoc                  0.6.0                    py27_0  
odo                       0.5.0                    py27_1  
olefile                   0.44                     py27_0  
openpyxl                  2.4.7                    py27_0  
openssl                   1.0.2l                        0  
packaging                 16.8                     py27_0  
pandas                    0.20.1              np112py27_0  
pandocfilters             1.4.1                    py27_0  
pango                     1.40.3                        1  
partd                     0.3.8                    py27_0  
patchelf                  0.9                           0  
path.py                   10.3.1                   py27_0  
pathlib2                  2.2.1                    py27_0  
patsy                     0.4.1                    py27_0  
pcre                      8.39                          1  
pep8                      1.7.0                    py27_0  
pexpect                   4.2.1                    py27_0  
pickleshare               0.7.4                    py27_0  
pillow                    4.1.1                    py27_0  
pip                       9.0.1                    py27_1  
pixman                    0.34.0                        0  
pkginfo                   1.4.1                    py27_0  
ply                       3.10                     py27_0  
prompt_toolkit            1.0.14                   py27_0  
psutil                    5.2.2                    py27_0  
ptyprocess                0.5.1                    py27_0  
py                        1.4.33                   py27_0  
pyasn1                    0.2.3                    py27_0  
pycairo                   1.10.0                   py27_0  
pycosat                   0.6.2                    py27_0  
pycparser                 2.17                     py27_0  
pycrypto                  2.6.1                    py27_6  
pycurl                    7.43.0                   py27_2  
pyflakes                  1.5.0                    py27_0  
pygments                  2.2.0                    py27_0  
pylint                    1.6.4                     <pip>
pylint                    1.6.4                    py27_1  
pyodbc                    4.0.16                   py27_0  
pyopenssl                 17.0.0                   py27_0  
pyparsing                 2.1.4                    py27_0  
pyqt                      5.6.0                    py27_2  
pytables                  3.3.0               np112py27_0  
pytest                    3.0.7                    py27_0  
python                    2.7.13                        0  
python-dateutil           2.6.0                    py27_0  
pytz                      2017.2                   py27_0  
pywavelets                0.5.2               np112py27_0  
pyyaml                    3.12                     py27_0  
pyzmq                     16.0.2                   py27_0  
qt                        5.6.2                         4  
qtawesome                 0.4.4                    py27_0  
qtconsole                 4.3.0                    py27_0  
qtpy                      1.2.1                    py27_0  
readline                  6.2                           2  
redis                     3.2.0                         0  
redis-py                  2.10.5                   py27_0  
requests                  2.14.2                   py27_0  
rope                      0.9.4                    py27_1  
ruamel_yaml               0.11.14                  py27_1  
scandir                   1.5                      py27_0  
scikit-image              0.13.0              np112py27_0  
scikit-learn              0.18.1              np112py27_1  
scipy                     0.19.0              np112py27_0  
seaborn                   0.7.1                    py27_0  
setuptools                27.2.0                   py27_0  
simplegeneric             0.8.1                    py27_1  
singledispatch            3.4.0.3                  py27_0  
sip                       4.18                     py27_0  
six                       1.10.0                   py27_0  
snowballstemmer           1.2.1                    py27_0  
sockjs-tornado            1.0.3                    py27_0  
sortedcollections         0.5.3                    py27_0  
sortedcontainers          1.5.7                    py27_0  
sphinx                    1.5.6                    py27_0  
sphinx_rtd_theme          0.2.4                    py27_0  
spyder                    3.1.4                    py27_0  
sqlalchemy                1.1.9                    py27_0  
sqlite                    3.13.0                        0  
ssl_match_hostname        3.4.0.2                  py27_1  
statsmodels               0.8.0               np112py27_0  
subprocess32              3.2.7                    py27_0  
sympy                     1.0                      py27_0  
tblib                     1.3.2                    py27_0  
terminado                 0.6                      py27_0  
testpath                  0.3                      py27_0  
tk                        8.5.18                        0  
toolz                     0.8.2                    py27_0  
tornado                   4.5.1                    py27_0  
traitlets                 4.3.2                    py27_0  
unicodecsv                0.14.1                   py27_0  
unixodbc                  2.3.4                         0  
util-linux                2.21                          0  
wcwidth                   0.1.7                    py27_0  
werkzeug                  0.12.2                   py27_0  
wheel                     0.29.0                   py27_0  
widgetsnbextension        2.0.0                    py27_0  
wrapt                     1.10.8                    <pip>
wrapt                     1.10.10                  py27_0  
xlrd                      1.0.0                    py27_0  
xlsxwriter                0.9.6                    py27_0  
xlwt                      1.2.0                    py27_0  
xz                        5.2.2                         1  
yaml                      0.1.6                         0  
zeromq                    4.1.5                         0  
zict                      0.1.2                    py27_0  
zlib                      1.2.8                         3  
--------------------------------------------------------------------------------
If requested, please copy and paste the information between
the dashed (----) lines, or from a given specific section as
appropriate.

=============================================================
IMPORTANT: Please ensure that you are happy with sharing the
contents of the information present, any information that you
wish to keep private you should remove before sharing.
=============================================================

```

**Comment by seibert** [comment]
OK, interesting to see you are using the CUDA 9 pre-release and libNVVM is not being detected.  I would have assumed Numba would raise an exception rather try to run.  I'll see if we can try out CUDA 9 on a test machine and figure out if there is some problem we're not aware of yet.

**Comment by seibert** [comment]
As a side note: `cuda.reduce` was rewritten for Numba 0.34 to improve performance.  I'm curious if it fails the same way for you.

**Comment by MikeQUS** [comment]
well, I tried it on a different machine with CUDA 7 and I got the wrong answer, same as the one in the question
Here is the spec of the other machine:

```
__Hardware Information__
Machine               : x86_64
CPU Name              : ivybridge
CPU Features          : 
aes avx cmov cx16 f16c fsgsbase mmx pclmul popcnt rdrnd sse sse2 sse3 sse4.1
sse4.2 ssse3 xsave xsaveopt

__OS Information__
Platform              : Darwin-14.5.0-x86_64-i386-64bit
Release               : 14.5.0
System Name           : Darwin
Version               : Darwin Kernel Version 14.5.0: Tue Apr 11 16:12:42 PDT 2017; root:xnu-2782.50.9.2.3~1/RELEASE_X86_64
OS specific info      : 10.10.5   x86_64

__Python Information__
Python Compiler       : GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)
Python Implementation : CPython
Python Version        : 3.5.3
Python Locale         : en_US UTF-8

__LLVM information__
LLVM version          : 4.0.0

__CUDA Information__
Found 1 CUDA devices
id 0      b'GeForce GT 650M'                              [SUPPORTED]
                      compute capability: 3.0
                           pci device id: 0
                              pci bus id: 1
Summary:
	1/1 devices are supported
CUDA driver version   : 7050
CUDA libraries:
Finding cublas
	named  libcublas.7.5.dylib
	trying to open library...	ok
Finding cusparse
	named  libcusparse.7.5.dylib
	trying to open library...	ok
Finding cufft
	named  libcufft.7.5.dylib
	trying to open library...	ok
Finding curand
	named  libcurand.7.5.dylib
	trying to open library...	ok
Finding nvvm
	named  libnvvm.3.0.0.dylib
	trying to open library...	ok
	finding libdevice for compute_20...	ok
	finding libdevice for compute_30...	ok
	finding libdevice for compute_35...	ok
	finding libdevice for compute_50...	ok

__Conda Information__
conda_build_version   : 1.21.3
conda_env_version     : 4.3.23
platform              : osx-64
python_version        : 3.5.3.final.0
root_writable         : True

__Current Conda Env__
_license                  1.1                      py35_1  
_nb_ext_conf              0.2.0                    py35_0  
alabaster                 0.7.10                   py35_0  
anaconda                  4.4.0               np112py35_0  
anaconda-client           1.6.3                    py35_0  
anaconda-navigator        1.6.2                    py35_0  
anaconda-project          0.6.0                    py35_0  
appnope                   0.1.0                    py35_0  
appscript                 1.0.1                    py35_0  
argcomplete               1.0.0                    py35_1  
asn1crypto                0.22.0                   py35_0  
astroid                   1.4.9                    py35_0  
astropy                   1.3.2               np112py35_0  
asv                       0.3.dev1059+47ae6767           <pip>
asv                       0.2.dev1053+011815d2           <pip>
babel                     2.4.0                    py35_0  
backports                 1.0                      py35_0  
beautifulsoup4            4.6.0                    py35_0  
bitarray                  0.8.1                    py35_0  
blaze                     0.10.1                   py35_0  
bleach                    1.5.0                    py35_0  
bokeh                     0.12.5                   py35_1  
boto                      2.46.1                   py35_0  
bottleneck                1.2.1               np112py35_0  
cffi                      1.10.0                   py35_0  
chardet                   3.0.3                    py35_0  
chest                     0.2.3                    py35_0  
click                     6.7                      py35_0  
cloudpickle               0.2.2                    py35_0  
clyent                    1.2.2                    py35_0  
colorama                  0.3.9                    py35_0  
conda                     4.3.23                   py35_0  
conda-build               1.21.3                   py35_0  
conda-env                 2.6.0                         0  
configobj                 5.0.6                    py35_0  
contextlib2               0.5.5                    py35_0  
cryptography              1.8.1                    py35_0  
cudatoolkit               7.5                           0  
curl                      7.52.1                        0  
cycler                    0.10.0                   py35_0  
cython                    0.25.2                   py35_0  
cytoolz                   0.8.2                    py35_0  
dask                      0.14.3                   py35_1  
datashape                 0.5.4                    py35_0  
decorator                 4.0.11                   py35_0  
dill                      0.2.5                    py35_0  
distributed               1.16.3                   py35_0  
docutils                  0.13.1                   py35_0  
dynd-python               0.7.2                    py35_0  
entrypoints               0.2.2                    py35_1  
et_xmlfile                1.0.1                    py35_0  
fastcache                 1.0.2                    py35_1  
flask                     0.12.2                   py35_0  
flask-cors                3.0.2                    py35_0  
freetype                  2.5.5                         2  
get_terminal_size         1.0.0                    py35_0  
gevent                    1.2.1                    py35_0  
greenlet                  0.4.12                   py35_0  
h5py                      2.7.0               np112py35_0  
hdf5                      1.8.17                        1  
heapdict                  1.0.0                    py35_1  
html5lib                  0.999                    py35_0  
icu                       54.1                          0  
idna                      2.5                      py35_0  
imagesize                 0.7.1                    py35_0  
ipykernel                 4.6.1                    py35_0  
ipython                   5.3.0                    py35_0  
ipython_genutils          0.2.0                    py35_0  
ipywidgets                6.0.0                    py35_0  
isort                     4.2.5                    py35_0  
itsdangerous              0.24                     py35_0  
jbig                      2.1                           0  
jdcal                     1.3                      py35_0  
jedi                      0.10.2                   py35_2  
jinja2                    2.9.6                    py35_0  
jpeg                      9b                            0  
jsonschema                2.6.0                    py35_0  
jupyter                   1.0.0                    py35_3  
jupyter_client            5.0.1                    py35_0  
jupyter_console           5.1.0                    py35_0  
jupyter_core              4.3.0                    py35_0  
lazy-object-proxy         1.2.2                    py35_0  
libdynd                   0.7.2                         0  
libiconv                  1.14                          0  
libpng                    1.6.27                        0  
libtiff                   4.0.6                         3  
libxml2                   2.9.4                         0  
libxslt                   1.1.29                        0  
llvmlite                  0.18.0                   py35_0  
locket                    0.2.0                    py35_1  
lxml                      3.7.3                    py35_0  
markupsafe                0.23                     py35_2  
matplotlib                2.0.2               np112py35_0  
mistune                   0.7.4                    py35_0  
mkl                       2017.0.1                      0  
mkl-service               1.1.2                    py35_3  
mpmath                    0.19                     py35_1  
msgpack-python            0.4.8                    py35_0  
multipledispatch          0.4.9                    py35_0  
navigator-updater         0.1.0                    py35_0  
nb_anacondacloud          1.1.0                    py35_0  
nb_conda                  1.1.0                    py35_0  
nb_conda_kernels          1.0.3                    py35_0  
nbconvert                 5.1.1                    py35_0  
nbformat                  4.3.0                    py35_0  
nbpresent                 3.0.2                    py35_0  
networkx                  1.11                     py35_0  
nltk                      3.2.3                    py35_0  
nose                      1.3.7                    py35_1  
notebook                  5.0.0                    py35_0  
numba                     0.33.0              np112py35_0  
numexpr                   2.6.2               np112py35_0  
numpy                     1.12.1                   py35_0  
numpydoc                  0.6.0                    py35_0  
odo                       0.5.0                    py35_1  
olefile                   0.44                     py35_0  
openpyxl                  2.4.7                    py35_0  
openssl                   1.0.2l                        0  
packaging                 16.8                     py35_0  
pandas                    0.20.1              np112py35_0  
pandocfilters             1.4.1                    py35_0  
partd                     0.3.8                    py35_0  
path.py                   10.3.1                   py35_0  
pathlib2                  2.2.1                    py35_0  
patsy                     0.4.1                    py35_0  
pep8                      1.7.0                    py35_0  
pexpect                   4.2.1                    py35_0  
pickleshare               0.7.4                    py35_0  
pillow                    4.1.1                    py35_0  
pip                       9.0.1                    py35_1  
pip                       9.0.0                     <pip>
ply                       3.10                     py35_0  
prompt_toolkit            1.0.14                   py35_0  
psutil                    5.2.2                    py35_0  
ptyprocess                0.5.1                    py35_0  
py                        1.4.33                   py35_0  
pyasn1                    0.1.9                    py35_0  
pycosat                   0.6.2                    py35_0  
pycparser                 2.17                     py35_0  
pycrypto                  2.6.1                    py35_6  
pycurl                    7.43.0                   py35_2  
pyflakes                  1.5.0                    py35_0  
pygments                  2.2.0                    py35_0  
pylint                    1.6.4                    py35_1  
pyodbc                    4.0.16                   py35_0  
pyopenssl                 17.0.0                   py35_0  
pyparsing                 2.1.4                    py35_0  
pyqt                      5.6.0                    py35_1  
pytables                  3.3.0               np112py35_0  
pytest                    3.0.7                    py35_0  
python                    3.5.3                         1  
python-dateutil           2.6.0                    py35_0  
python.app                1.2                      py35_4  
pytz                      2017.2                   py35_0  
pywavelets                0.5.2               np112py35_0  
pyyaml                    3.12                     py35_0  
pyzmq                     16.0.2                   py35_0  
qt                        5.6.2                         2  
qtawesome                 0.4.4                    py35_0  
qtconsole                 4.3.0                    py35_0  
qtpy                      1.2.1                    py35_0  
readline                  6.2                           2  
redis                     3.2.0                         0  
redis-py                  2.10.5                   py35_0  
requests                  2.14.2                   py35_0  
rope                      0.9.4                    py35_1  
ruamel_yaml               0.11.14                  py35_1  
scikit-image              0.13.0              np112py35_0  
scikit-learn              0.18.1              np112py35_1  
scipy                     0.19.0              np112py35_0  
seaborn                   0.7.1                    py35_0  
setuptools                27.2.0                   py35_0  
simplegeneric             0.8.1                    py35_1  
singledispatch            3.4.0.3                  py35_0  
sip                       4.18                     py35_0  
six                       1.10.0                   py35_0  
snowballstemmer           1.2.1                    py35_0  
sockjs-tornado            1.0.3                    py35_0  
sortedcollections         0.5.3                    py35_0  
sortedcontainers          1.5.7                    py35_0  
sphinx                    1.5.6                    py35_0  
sphinx_rtd_theme          0.1.9                    py35_0  
spyder                    3.1.4                    py35_0  
sqlalchemy                1.1.9                    py35_0  
sqlite                    3.13.0                        0  
statsmodels               0.8.0               np112py35_0  
sympy                     1.0                      py35_0  
tabulate                  0.7.5                     <pip>
tblib                     1.3.2                    py35_0  
terminado                 0.6                      py35_0  
testpath                  0.3                      py35_0  
tk                        8.5.18                        0  
toolz                     0.8.2                    py35_0  
tornado                   4.5.1                    py35_0  
traitlets                 4.3.2                    py35_0  
unicodecsv                0.14.1                   py35_0  
unixodbc                  2.3.4                         0  
wcwidth                   0.1.7                    py35_0  
werkzeug                  0.12.2                   py35_0  
wheel                     0.29.0                   py35_0  
widgetsnbextension        2.0.0                    py35_0  
wrapt                     1.10.10                  py35_0  
xlrd                      1.0.0                    py35_0  
xlsxwriter                0.9.6                    py35_0  
xlwings                   0.10.4                   py35_0  
xlwt                      1.2.0                    py35_0  
xz                        5.2.2                         1  
yaml                      0.1.6                         0  
zict                      0.1.2                    py35_0  
zlib                      1.2.8                         3  
--------------------------------------------------------------------------------
If requested, please copy and paste the information between
the dashed (----) lines, or from a given specific section as
appropriate.

=============================================================
IMPORTANT: Please ensure that you are happy with sharing the
contents of the information present, any information that you
wish to keep private you should remove before sharing.
=============================================================

```

**Comment by seibert** [comment]
OK, I can confirm your bug when I run with Numba 0.33.0.  This was fixed in Numba 0.34.0.

**Comment by MikeQUS** [comment]
Interestingly, I tried the `stream` functionality on Java to leverage parallel `reduce` and got the same answer as CUDA but different than sequential reduce

```java

int n = 10;
float inputArray[] = new float[n];
ArrayList<Float> inputList = new ArrayList<Float>();
for (int i=0; i<n; i++)
{
    inputArray[i] = i+1;
    inputList.add(inputArray[i]);
}
Optional<Float> resultStream = inputList.stream().parallel().reduce((x, y) -> x+y*20);
float resultCPU = array[0];
for (int i = 1; i < array.length; i++)
{
    resultCPU = resultCPU + array[i] * 20;            
}
System.out.println("CPU "+resultCPU); // CPU 10541.0
System.out.println("Stream "+resultStream.get()); // Stream 1.2466232E8

```

**Comment by MikeQUS** [comment]
I installed Numba 0.34.0 and you are right results are correct for the example I gave in the question.
Although, when you make the array larger, the problem reappear:

```python

import numpy
from numba import cuda
from functools import reduce

A = (numpy.arange(100, dtype=numpy.float64)) + 1
cuda.reduce(lambda a, b: a + b * 20)(A) 
# result 12952749821.0
reduce(lambda a, b: a + b * 20, A) 
# result 100981.0

import numba
numba.__version__
# '0.34.0+5.g1762237'

```

**Comment by seibert** [comment]
@sklam: Can you take a look at this?

**Comment by MikeQUS** [comment]
@seibert  Thanks by the way for your fast response and help
I would like to remove my systems spec and just leave the parts you were interested in

**Comment by seibert** [comment]
@sklam just pointed out to me that parallel reductions require a commutative reduction function.  Your lambda function is not commutative:
```
lambda a, b: a + b * 20
```
This also explains why you see the same strange behavior in Java, which is probably doing a tree-based reduction similar to the GPU.  The rewrite of `cuda.reduce` between 0.33 and 0.34 was a red herring as it just changed the order of operations and accidentally made the size 10 case work, as you noted.

So, basically, for this kernel, parallel reduction can't work in any system.

**Comment by MikeQUS** [comment]
Make sense, thank you so much both of you @seibert @sklam  for clarifying this to me.
It would be very interesting to find out how to identify commutative reduction within the compiler and throw and exception in those cases.
